source/http_stream.c

```diff
@@ -51,148 +50,163 @@ struct aws_http_stream *aws_py_get_http_stream(PyObject *stream) {
-        stream->on_incoming_headers_received, "(OiO)", stream->received_headers, response_code, has_body_obj);
+    /* Deliver the built up list of (name,value) tuples */
+    PyObject *result =
+        PyObject_CallMethod(stream->self_proxy, "_on_response", "(iO)", response_code, stream->received_headers);
     if (!result) {
-        PyErr_WriteUnraisable(PyErr_Occurred());
+        aws_result = aws_raise_py_error();
+        goto done;
+    }
+    Py_DECREF(result);
+
+    /* Clear the list so we're ready for next header block */
+    result = PyObject_CallMethod(stream->received_headers, "clear", "()", NULL);
+    if (!result) {
+        aws_result = aws_raise_py_error();
+        goto done;
     }
-    Py_XDECREF(result);
-    Py_CLEAR(stream->received_headers);
-    Py_CLEAR(stream->on_incoming_headers_received);
+    Py_DECREF(result);
+
+done:
     PyGILState_Release(state);
+    /*************** GIL RELEASE ***************/
+
+    return aws_result;
 }
 
-static void s_on_incoming_response_body(
-    struct aws_http_stream *internal_stream,
+static int s_on_incoming_body(
+    struct aws_http_stream *native_stream,
     const struct aws_byte_cursor *data,
-    size_t *out_window_update_size,
     void *user_data) {
-    (void)internal_stream;
-    (void)out_window_update_size;
+
+    (void)native_stream;
 
     struct http_stream_binding *stream = user_data;
 
+    if (data->len > PY_SSIZE_T_MAX) {
+        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    }
+    Py_ssize_t data_len = (Py_ssize_t)data->len;
+
+    int aws_result = AWS_OP_SUCCESS;
+
+    /*************** GIL ACQUIRE ***************/
     PyGILState_STATE state = PyGILState_Ensure();
 
-    Py_ssize_t data_len = (Py_ssize_t)data->len;
-    PyObject *result =
-        PyObject_CallFunction(stream->on_incoming_body, "(" BYTE_BUF_FORMAT_STR ")", (const char *)data->ptr, data_len);
+    PyObject *result = PyObject_CallMethod(
+        stream->self_proxy, "_on_body", "(" READABLE_BYTES_FORMAT_STR ")", (const char *)data->ptr, data_len);
     if (!result) {
-        PyErr_WriteUnraisable(PyErr_Occurred());
+        aws_result = aws_raise_py_error();
+        goto done;
     }
-    Py_XDECREF(result);
+    Py_DECREF(result);
+
+done:
     PyGILState_Release(state);
+    /*************** GIL RELEASE ***************/
+
+    return aws_result;
 }
 
-static void s_on_stream_complete(struct aws_http_stream *internal_stream, int error_code, void *user_data) {
-    (void)internal_stream;
+static void s_on_stream_complete(struct aws_http_stream *native_stream, int error_code, void *user_data) {
+    (void)native_stream;
     struct http_stream_binding *stream = user_data;
 
+    /*************** GIL ACQUIRE ***************/
     PyGILState_STATE state = PyGILState_Ensure();
 
-    PyObject *result = PyObject_CallFunction(stream->on_stream_completed, "(i)", error_code);
+    /* Set this just in case callback fires before aws_http_connection_make_request() has even returned */
```

Comment:
If that's a possibility, we have unleashed Zalgo and we should not do that.

Reply:
This is that thing we chatted about where it seems like we might want some kind of `.activate()` function on the stream that users need to call it kick it off. I tried really hard to do things RAII in aws-c-http, not having objects in a "half-initialized" state, but we end up in situations like this...

If we're going to address this in aws-c-http, we need to set aside time to do so.

Reply:
Adding a task to address this. Neither of us like the `stream.activate()` solution, but neither of us can think of anything else bulletproof